auth.py

```diff
@@ -0,0 +1,304 @@
+
+        with open(os.path.join(os.path.dirname(__file__), 
+                "saml-templates/sp-adfs-stsBinaryToken.xml"), "r") as file:
+            saml = file.read()
+
+        # Format saml to receive BinarySecurityToken
+        saml = saml.format(customSTSAssertion=samlAssertion,
+                           msoEndpoint="sharepoint.com")
+        response = requests.post(url=MSO_AUTH_URL, data=saml, headers=headers)
+        # Parse and extract token from returned XML
+        try:
+            root = et.fromstring(response.text)
+        except et.ParseError:
+            print("BinarySecurityToken request failed. The server did not send a valid response\r", end="")
+            return
+        try:
+            binarySecurityToken = (re.search('BinarySecurityToken Id.*>([^<]+)', response.text)).group(1)
+        except AttributeError:
+            print("BinarySecurityToken requested, but response did not include the STS SAML Assertion.\r", end="")
+            return
+        
+        print("BinarySecurityToken received.")
+        self.token = binarySecurityToken
+
+        return True
+    
+    def _get_cookie(self):
+        # regex will get the 'tenant url' from the site provided for login
+        SPO_IDCRL_URL = "https://" + re.search('(.+\.com)/', self.site).group(1) + "/_vti_bin/idcrl.svc/"
+        
+        # HTTP Post request with authorization headers
+        headers = {"Authorization": "BPOSIDCRL " + self.token, 
+                   "X-IDCRL_ACCEPTED":"t", 
+                   "User-Agent": "" }
+        response = requests.get(url=SPO_IDCRL_URL, headers=headers)
+
+        if response.status_code == requests.codes.ok:
+            print("ADFS Authentication successful")
+
+            # Add the SPOIDCRL cookie to the session
+            self.cookie = self._buildcookie(response.cookies)
+            
+            
+            return True
+        else:
+            print("ADFS Authentication failed")
+
+    def _get_digest(self):
+        """Check and refresh sites cookie and request digest"""
+        
+        if self.expire <= datetime.now():
+            # Template for SOAP digest request using <tenant>.sharepoint.com/sites/<sub site>/_vti_bin/sites
+            headers = ({"SOAPAction": "http://schemas.microsoft.com/sharepoint/soap/GetUpdatedFormDigestInformation",
+                                "Host" : self.tenantBaseURL,
+                                "Content-Type" : "text/xml;charset=utf-8",
+                                "Cookie" : self.cookie,
+                                "X-RequestForceAuthentication" : "true"})
+            digestEnvelope = '<?xml version="1.0" encoding="utf-8"?><soap:Envelope xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:soap="http://schemas.xmlsoap.org/soap/envelope/"><soap:Body><GetUpdatedFormDigest xmlns="http://schemas.microsoft.com/sharepoint/soap/" /></soap:Body></soap:Envelope>'
```

Comment:
Can this XML be loaded from a template file like the others?

Reply:
✔️ 
added `saml-templates/sp-updateDigest.xml` with the SOAP envelop to request a new digest